### DISCUSSION TITLE
Coroutine scheduling using libuv?

### DISCUSSION THREAD

**Discussion by cmm** [libuv/libuv#4110] [Q&A] [answered]
I'm playing with implementing a simple coroutine library backed by libuv.  Most async `uv` APIs lend themselves nicely to the cheerful stupid pattern where you make an async call while arranging for the callback to resume the coroutine, then yield.

But things get trickier with the `uv_<foo>_{stop,start}()`-type APIs, where a callback may fire multiple times.  A simple band-aid is to `uv_<foo>_stop()` first thing in the callback, but I wonder if that may result in missed activity?

Like, stopping/restarting a stream read should be safe (right?), but I guess polling isn't?  What about UDP recv?

(Any pointers to working prior art are as welcome as straight answers!)

**Comment by bnoordhuis** [comment, marked as answer]
Calling uv_thing_stop() should always be safe, at least from a libuv perspective.

I suppose polling some hypothetical edge-triggered resource could lead to missed events but I can't think of any real-world resource that works that way. Maybe stuff under /dev or /proc?

**Reply by cmm** [reply]
So as long as the resource being polled is level-triggered all is good, gotcha!

What about `uv_udp_recv_stop()` though?  From my looking at the code it seems that the handle gets "stopped" but the socket therein stays bound (which would be what I'm hoping for), but I'm not completely sure.

And what about `uv_fs_{event,poll}_stop()`?

(Thankfully I don't care about `uv_signal_*()`, so that's good)

**Reply by bnoordhuis** [reply]
> the handle gets "stopped" but the socket therein stays bound

Yes, that's right.

> And what about uv_fs_{event,poll}_stop()?

It's complicated and platform dependent. :-)

**Comment by cmm** [comment]
Is there any way to "stop" `uv_listen()`?
(And, while we are at it, are there any other such "watcher"-type-sans-"stop" APIs that I missed?)

**Reply by bnoordhuis** [reply]
Not at the moment. I have plans of making accepting connections and reading data from them request-based (like writing already is) but no ETA. The Windows part keeps stumping me. :-/

**Reply by winterland1989** [reply]
I think what we really want is an only called once per `uv_run(UV_RUN_ONCE)` guarantee, which can be solved by adding a backlog queue and returning accepted FDs all together. This is currently how Haskell binding works.

We also initialize a `uv_tcp_t` from FD on possibly another `uv_loop`(possibly running on another core), that's how work distribution works. Providing initialized uv structs in callback stopped us from doing so.

**Comment by haoyu234** [comment]
Maybe we can add some `uv_<foo>_once()`-type APIs, such as `uv_read_once`, `uv_listen_once`

**Reply by bnoordhuis** [reply]
That's something we've been wanting to do for a long time but it's Complicated.

https://github.com/bnoordhuis/libuv/commit/08beaedf7f7c6302da11a8b99ec663668f861df3 tackles just `uv_accept()` and is already a 1,000+ lines diff.

**Comment by snej** [comment]
I've been working on a C++20 coroutine runtime using libuv. I've got most functionality running, plus HTTP requests (using llhttp) and WebSockets. The code's not very mature yet, though; all I have are a minimal set of unit tests and two very simple CLI tools. I plan to announce it soon when it's a bit more solid. But reply here if you want to hear more!
